Add "File type" filter Chip

lookacat opened the pull request:
<!--
Thanks for submitting a change to ownCloud!

This is the bug tracker for Web. Find other components at https://github.com/owncloud/core/blob/master/.github/CONTRIBUTING.md#guidelines

For fixing potential security issues please see https://owncloud.org/security/

To make it possible for us to get your change reviewed and merged please carefully fill out the requested information below.

Please note that any kind of change needs first be submitted to the master branch which holds the next major version of ownCloud.

We will carefully discuss if your change can or has to be backported to stable branches.

Please set the following labels:

- Set label "Status:Needs-Review" for review or "Status:In-Progress" in case the PR still has open tasks
- Set label "Category:*" where it fits best
- Assignment: assign to self
- Reviewers: pick at least one
-->

## Description
**The capability for mimeTypes needs to be enabled OR find** `unref(modifiedDateCapability).enabled` **and add a** `!`
**Backend is not on master yet, needs Romans PR (https://github.com/owncloud/ocis/pull/7602) as backend**
See https://github.com/owncloud/web/issues/9780

## Related Issue
<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->
- Fixes https://github.com/owncloud/web/issues/9780

## Screenshots (if appropriate):
![Screenshot](https://github.com/owncloud/web/assets/786551/8603d9c0-f297-4e7e-9971-5c02dc46b90a)

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests

## Checklist:
<!-- Tick the checkboxes when done. -->
<!-- Raise documentation ticket in https://github.com/owncloud/documentation -->
- [x] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 

## Open tasks:
<!-- In case of incomplete PR, please list the open tasks here -->
- [ ] https://github.com/owncloud/ocis/pull/7602 needs to be used for the backend OR its merged

JanAckermann reviewed (changes requested):
Nice, I'd love to have some tests if possible

JammingBen commented on packages/web-app-files/src/components/Search/List.vue:
```diff
@@ -7,6 +7,26 @@
+          filter-name="mimeTypes"
+        >
+          <template #image="{ item }">
+            <div class="tag-option-wrapper oc-flex oc-flex-middle">
```
```suggestion
            <div class="type-option-wrapper oc-flex oc-flex-middle">
```

JammingBen commented on packages/web-app-files/src/components/Search/List.vue:
```diff
@@ -287,6 +322,33 @@ export default defineComponent({
+      audio: { label: $gettext('Audio'), icon: 'mp3' },
+      archive: { label: $gettext('Archive'), icon: 'zip' }
+    }
+    const availableMimeTypeValues = []
```
```suggestion
    const availableMimeTypeValues: MimeTypeKeyword[] = []
```

JammingBen commented on packages/web-app-files/src/components/Search/List.vue:
```diff
@@ -246,12 +275,18 @@ export default defineComponent({
 
     const displayFilter = computed(() => {
-      return unref(fullTextSearchEnabled) || unref(availableTags).length
+      return (
+        unref(fullTextSearchEnabled) ||
+        unref(availableTags).length ||
+        unref(modifiedDateCapability).enabled
+      )
```
Shouldn't it also check for the `mimeTypeCapability`?

lookacat replied:
wasnt sure because it should be always there but more checks are better ^^

JammingBen reviewed (commented):
Please also add `q_mimeTypes` here to ensure that the filter stays active after closing an editor on the search result page:

https://github.com/owncloud/web/blob/7e5340ad822a3d84a364455452f1ed7a83515232/packages/web-pkg/src/router/common.ts#L31

lookacat commented:
I addressed most issues, i didn't rename the capability yet because it would break the whole thing. I'll have to wait until the linked PR (in the description) from Roman is changed.
Tests will follow

lookacat commented:
e2e tests added, but will only work if romans PR is on master

JammingBen commented on tests/e2e/cucumber/features/smoke/search.feature:
```diff
@@ -171,3 +171,36 @@ Feature: Search
+    And "Alice" selects mediaType "Archive" from the search result filter chip
+    Then following resources should be displayed in the files list for user "Alice"
+      | resource      |
+      | mediaTest.zip |
```
```suggestion
      | mediaTest.zip |
    And "Alice" logs out
```

JammingBen reviewed (approved):
LGTM 👍

kulmann reviewed (commented):
Found one weird issue: 
- open file type filter
- select "Document" and "Folder"
- use text filter in the dropdown with whatever string and delete it again
- selected items (Document and Folder) are re-arranged to the top of the list

Fine to solve this in a followup if this is an issue of the filter dropdown (that's what I suspect). Please open a bug issue after investigation.

kulmann commented on packages/web-app-files/src/components/Search/List.vue:
```diff
@@ -287,6 +325,33 @@ export default defineComponent({
+        id: key,
+        ...mediaTypeMapping[key]
+      })
+    })
```
Could you please apply an order? I'd suggest:
- Folder
- File
- all other options sorted lexicographically

I'd put `Folder` and `File` to the top because they're somewhat special.

kulmann replied:
Sorry, my bad. The order is specified in the issue and should come from the backend. You did that correctly. 👍

lookacat replied:
alright 👍🏼

lookacat commented:
@kulmann please rereview, followup issue is created and im already on it https://github.com/owncloud/web/issues/9938

kulmann reviewed (approved):
Love it! 😍
